pkg/vm/vm.go

```diff
@@ -162,6 +168,26 @@ func (vm *VirtualMachine) RunInstruction(instruction *Instruction) error {
 	return nil
 }
 
+// It returns the current trace entry, the public memory, and the occurrence of an error
+func (vm *VirtualMachine) Proof() ([]Context, []*f.Element, error) {
+	if !vm.config.ProofMode {
+		return nil, nil, fmt.Errorf("cannot get proof if proof mode is off")
+	}
+
+	// update pc, fp and ap to be their real value
+	// that is, pc_i + 0, ap + len(programBytecode), fp + len(programBytecode)
+	totalBytecode := vm.MemoryManager.Memory.Segments[ProgramSegment].Len()
+	for i := range vm.Trace {
+		vm.Trace[i].Ap += totalBytecode
```

Comment:
So you suggest to convert them to address and then use relocation logic, I don't think is zero cost because the of the extra heap allocations. What if I made a`Relocate` function for the `Context` datatype. It would look like:

```go
for i:= range vm.Trace {
    vm.Trace[i].Relocate()
}
```

Reply:
Yeah, could you make a func for ```Context``` relocating then please? It just feels like it wouldn't hurt if that was abstracted